binaries/opae.io/opae/io/pci.py

```diff
@@ -0,0 +1,167 @@
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISEDL OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+import re
+from pathlib import Path
+from collections import namedtuple
+
+
+PCI_ADDRESS_PATTERN = (r'(?P<pci_address>'
+                       r'(?:(?P<segment>[\da-f]{4}):)?'
+                       r'(?P<bdf>(?P<bus>[\da-f]{2}):'
+                       r'(?P<device>[\da-f]{2})\.(?P<function>\d)))')
+PCI_ADDRESS_RE = re.compile(PCI_ADDRESS_PATTERN, re.IGNORECASE)
+
+
+VENDOR_DEVICE_PATTERN = r'(?P<vendor>[\da-f]{0,4}):(?P<device>[\da-f]{0,4})'
+VENDOR_DEVICE_RE = re.compile(VENDOR_DEVICE_PATTERN, re.IGNORECASE)
+
+pci_id = namedtuple(
+    'pci_id', [
+        'vendor', 'device', 'subsystem_vendor', 'subsystem_device'], defaults=[
+            0, 0])
+pci_id.short = lambda i: pci_id(i.vendor, i.device)
+
+
+SYSFS_PCI_DEVICES = '/sys/bus/pci/devices'
+
+
+def pci_address(inp):
+    m = PCI_ADDRESS_RE.match(inp)
+    if not m:
+        raise ValueError('wrong pci address format: {}'.format(inp))
+
+    d = m.groupdict()
+    return '{}:{}'.format(d.get('segment') or '0000', d['bdf']).lower()
+
+
+def vendev(inp):
+    m = VENDOR_DEVICE_RE.match(inp)
+    if not m:
+        raise ValueError('wrong vendor/device format: {}'.format(inp))
+    return dict((k, int(v, 16)) for k, v in m.groupdict().items() if v)
+
+
+def vid_did_for_address(pci_addr):
+    path = Path('/sys/bus/pci/devices', pci_addr)
+    vid = path.joinpath('vendor').read_text().strip()
+    did = path.joinpath('device').read_text().strip()
+    return (vid, did)
+
+
+class sysfs_attr:
+    def __init__(self, parent, path: Path):
+        self.parent = parent
+        self.path = path
+
+    @property
+    def value(self):
+        with self.parent.path.joinpath(self.path).open('r') as fp:
+            return fp.read()
+
+    def __str__(self):
+        return self.value.strip()
+
+    def __repr__(self):
+        return f'{self.parent}.{self.path} = {self}'
+
+    def __int__(self):
+        return int(self.value.strip(), 0)
+
+    def __eq__(self, other):
+        args = [self.value.strip()]
+        if isinstance(other, int):
+            args.append(0)
+        return type(other)(*args) == other
+
+
+CLASSES = {
+    0x01: "Mass storage controller",
+    0x02: "Network controller",
+    0x03: "Display controller",
+    0x04: "Multimedia device",
+    0x05: "Memory controller",
+    0x06: "Bridge device",
+    0x07: "Simple communication controllers ",
+    0x08: "Base system peripherals ",
+    0x09: "Input devices ",
+    0x0A: "Docking stations ",
+    0x0B: "Processors ",
+    0x0C: "Serial bus controllers ",
+    0x0D: "Wireless controller",
+    0x0E: "Intelligent I/O controllers",
+    0x0F: "Satellite communication controllers",
+    0x10: "Encryption/Decryption controllers",
+    0x11: "Data acquisition and signal processing controllers",
+    0x12: "Processing accelerators",
+    0x13: "Non-Essential Instrumentation"
+}
```

Comment:
It seems like this would be defined somewhere like pciutils, but then again I'm extremely lazy (in the perl sense).

Reply:
Perhaps but I don't want to have another dependency if I can avoid it.